src/Type/Php/MinMaxFunctionReturnTypeExtension.php

```diff
@@ -128,6 +135,50 @@ private function processType(
 		return $resultType;
 	}
 
+	private function mergeRangeTypes(
+		Type $firstType,
+		Type $secondType,
+		string $functionName
+	): ?Type
+	{
+		$rangeType = $intType = null;
+
+		if (
+			$firstType instanceof IntegerRangeType
+			&& $secondType instanceof ConstantIntegerType
+		) {
+			$rangeType = $firstType;
+			$intType = $secondType;
+		}
+
+		if (
+			$firstType instanceof ConstantIntegerType
+			&& $secondType instanceof IntegerRangeType
+		) {
+			$rangeType = $secondType;
+			$intType = $firstType;
```

Comment:
I have an idea for complete reimplementation of this - you can rewrite `min($x, $y)` as  `$x < $y ? $x : $y`. What if we did this instead (pseudocode)?

```php
$scope->getType(new Ternary(
    new Smaller($args[0], $args[1]),
    $args[0],
    $args[1]
));
```

And handled it in TypeSpecifier around this area? https://github.com/phpstan/phpstan-src/blob/f7250dbe9b23415f02961edc5f34e3f084e2b659/src/Analyser/TypeSpecifier.php#L465-L555

You might even find out it already works. Additionally, fixing some additional cases to make these tests pass would improve engine not just for `min`/`max`...

Reply:
I see. Do you mean we would achieve this by rewriting the AST? If so, would this rewriting happen within a extension somewhere or as part of the phpstan-core logic?

Reply:
No rewriting necessary. The idea is that the entire code of the extension for `min()` should be:

```
return $scope->getType(new Ternary(
    new Smaller($args[0], $args[1]),
    $args[0],
    $args[1]
));
```

And the whole logic should be handled in the linked code in TypeSpecifier.

Of course it gets a little complicated for multiple arguments and `$arg–>unpack` but it might be achievable with BooleanAnd and BooleanOr.

Reply:
to keep this PR as is in case of problems with the different approach, I created a new PR based on your suggestion:

https://github.com/phpstan/phpstan-src/pull/627

Reply:
Yeah, I love the new PR, thanks :)